components/common/src/listen_ctl_addr.rs

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) 2018 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+use std::net::{IpAddr, Ipv4Addr, SocketAddr, SocketAddrV4};
+use std::result;
+use std::str::FromStr;
+
+use error::{Error, Result};
+use EnvConfig;
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
```

Comment:
I don't think it's a good idea to make `ListenCtlAddr` implement `Copy`. In general, copy types should be very small and not contain heap-allocated data which may have more significant performance implications. Since `SocketAddr` doesn't implement `Copy`, that's probably a good hint that a type wrapping it shouldn't either. This has the benefit of forcing the code that uses it to either explicitly call `clone`, which makes the performance issues more apparent and easier to track down later, or figuring out how to make appropriate use of references.

I'm assuming you derived copy here, because without it, you get errors like this:
```
error[E0382]: use of moved value: `sup_addr`                                                                                                                                                                
   --> components/hab/src/main.rs:895:56                                                                                                                                                                    
    |                                                                                                                                                                                                       
873 |     SrvClient::connect(&sup_addr.into(), &secret_key)                                                                                                                                                 
    |                         -------- value moved here                                                                                                                                                     
...                                                                                                                                                                                                         
895 |     ui.status(Status::Applying, format!("via peer {}", sup_addr))?;                                                                                                                                   
    |                                                        ^^^^^^^^ value used here after move                                                                                                            
    |                                                                                                                                                                                                       
    = note: move occurs because `sup_addr` has type `common::types::ListenCtlAddr`, which does not implement the `Copy` trait                                                                               
```
What's happening is that previously we could use the same `sup_addr` **reference** in two places because  `SrvClient::connect` takes a `&SocketAddr` parameter. However, now that we have wrapped `SocketAddr` in a `ListenCtlAddr`, we must call `into` to convert to a `SocketAddr` (and then take a reference), but because `Into<SocketAddr> for ListenCtlAddr`'s `into` method takes ownership of `self`, you get the above error.

Also note that, it's generally not necessary to implement `Into` since implementing `From` gives `Into` automatically:
> Library authors should not directly implement this trait, but should prefer implementing the From trait, which offers greater flexibility and provides an equivalent Into implementation for free, thanks to a blanket implementation in the standard library.

See [the docs](http://doc.rust-lang.org/1.31.0/std/convert/trait.Into.html#method.into) for more.

However, in this case, what we want is something else: we want to be able to use a `ListenCtlAddr` where we could previously provide a `&SocketAddr`. Fortunately, there's a type for just that purpose: [`AsRef`](http://doc.rust-lang.org/1.31.0/core/convert/trait.AsRef.html). If you add this implementation:
```rust
impl AsRef<SocketAddr> for ListenCtlAddr {
    fn as_ref(&self) -> &SocketAddr {
        &self.0
    }
}
```
Then you can resolve the above error with:
```rust
    SrvClient::connect(sup_addr.as_ref(), &secret_key)
```
However, you may instead want to change the interface of `SrvClient::connect` to accept a `&ListenCtlAddr` instead of a `&SocketAddr`. After all, one of the points of this wrapping type is enforcing correctness through the compiler, right?

In that case, the call to `SrvClient::connect` stays as it was:
```rust
    SrvClient::connect(&sup_addr, &secret_key)
```
but the code *in* `connect` that needs a `&SocketAddr` specifically (because it's a library external to our project) makes use of our `AsRef` implementation:
```rust
        let conn = TcpStream::connect(addr.as_ref())
```

Reply:
Agree with not deriving `Copy` here... that snuck by me initially 👍 

(Also agree with all the rest, too)